File: android/src/main/kotlin/com/revolut/sdk/bridge/RevolutPayModels.kt
package com.revolut.sdk.bridge

import com.revolut.revolutpay.ui.button.ButtonParams
import com.revolut.revolutpay.ui.button.VariantModes
import com.revolut.revolutpay.ui.promotional.PromoBannerParams
import com.revolut.revolutpay.model.Customer
import com.revolut.revolutpay.model.DateOfBirth
import com.revolut.revolutpay.model.Country

/**
 * Data models and enums for Revolut Pay SDK integration
 * This file contains all the necessary data structures that Flutter can use
 * to communicate with the native Revolut Pay SDK
 */

// Environment enum for SDK initialization
enum class RevolutEnvironment(val value: String) {
    MAIN("MAIN"),
    SANDBOX("SANDBOX")
}

// Button radius options
enum class ButtonRadius(val value: String) {
    SMALL("SMALL"),
    MEDIUM("MEDIUM"),
    LARGE("LARGE")
}

// Button size options
enum class ButtonSize(val value: String) {
    SMALL("SMALL"),
    MEDIUM("MEDIUM"),
    LARGE("LARGE")
}

// Box text options for button customization
enum class BoxText(val value: String) {
    NONE("NONE"),
    GET_CASHBACK_VALUE("GET_CASHBACK_VALUE"),
    GET_CASHBACK_PERCENTAGE("GET_CASHBACK_PERCENTAGE")
}

// Button variant options
enum class ButtonVariant(val value: String) {
    LIGHT("LIGHT"),
    DARK("DARK")
}

// Customer data structure
data class CustomerData(
    val name: String? = null,
    val email: String? = null,
    val phone: String? = null,
    val dateOfBirth: DateOfBirthData? = null,
    val country: CountryData? = null
)

// Date of birth data structure
data class DateOfBirthData(
    val day: Int,
    val month: Int,
    val year: Int
)

// Country data structure
data class CountryData(
    val value: String // ISO 3166 2-letter country code
)

// Button parameters data structure
data class ButtonParamsData(
    val radius: ButtonRadius = ButtonRadius.MEDIUM,
    val size: ButtonSize = ButtonSize.LARGE,
    val boxText: BoxText = BoxText.NONE,
    val boxTextCurrency: String? = null,
    val variantModes: VariantModesData? = null
)

// Variant modes data structure
data class VariantModesData(
    val darkTheme: ButtonVariant = ButtonVariant.DARK,
    val lightTheme: ButtonVariant = ButtonVariant.LIGHT
)

// Promotional banner parameters data structure
data class PromoBannerParamsData(
    // Add specific promotional banner parameters as needed
    val customParam: String? = null
)

// Order result callback data
data class OrderResultData(
    val success: Boolean,
    val orderId: String? = null,
    val error: String? = null,
    val cause: String? = null
)

// Controller creation result data
data class ControllerResultData(
    val controllerId: String,
    val success: Boolean
)

// Button creation result data
data class ButtonResultData(
    val buttonId: String,
    val success: Boolean
)

// Banner creation result data
data class BannerResultData(
    val bannerId: String,
    val success: Boolean
)

// Payment flow data
data class PaymentFlowData(
    val orderToken: String,
    val savePaymentMethodForMerchant: Boolean = false
)

// SDK initialization data
data class SdkInitData(
    val environment: RevolutEnvironment,
    val returnUri: String,
    val merchantPublicKey: String,
    val requestShipping: Boolean = false,
    val customer: CustomerData? = null
)

/**
 * Extension functions to convert between Flutter data and SDK objects
 */

fun CustomerData.toSdkCustomer(): Customer {
    return Customer(
        name = this.name,
        email = this.email,
        phone = this.phone,
        dateOfBirth = this.dateOfBirth?.toSdkDateOfBirth(),
        country = this.country?.toSdkCountry()
    )
}

fun DateOfBirthData.toSdkDateOfBirth(): DateOfBirth {
    return DateOfBirth(
        day = this.day,
        month = this.month,
        year = this.year
    )
}

fun CountryData.toSdkCountry(): Country {
    return Country(this.value)
}

fun ButtonParamsData.toSdkButtonParams(): ButtonParams {
    return ButtonParams.Builder()
        .radius(this.radius.toSdkRadius())
        .size(this.size.toSdkSize())
        .boxText(this.boxText.toSdkBoxText())
        .boxTextCurrency(this.boxTextCurrency)
        .variantModes(this.variantModes?.toSdkVariantModes())
        .build()
}

fun ButtonRadius.toSdkRadius(): ButtonParams.Radius {
    return when (this) {
        ButtonRadius.SMALL -> ButtonParams.Radius.SMALL
        ButtonRadius.LARGE -> ButtonParams.Radius.LARGE
        else -> ButtonParams.Radius.MEDIUM
    }
}

fun ButtonSize.toSdkSize(): ButtonParams.Size {
    return when (this) {
        ButtonSize.SMALL -> ButtonParams.Size.SMALL
        ButtonSize.MEDIUM -> ButtonParams.Size.MEDIUM
        else -> ButtonParams.Size.LARGE
    }
}

fun BoxText.toSdkBoxText(): ButtonParams.BoxText {
    return when (this) {
        BoxText.GET_CASHBACK_VALUE -> ButtonParams.BoxText.GET_CASHBACK_VALUE
        BoxText.GET_CASHBACK_PERCENTAGE -> ButtonParams.BoxText.GET_CASHBACK_PERCENTAGE
        else -> ButtonParams.BoxText.NONE
    }
}

fun VariantModesData.toSdkVariantModes(): VariantModes {
    return VariantModes(
        darkTheme = this.darkTheme.toSdkVariant(),
        lightTheme = this.lightTheme.toSdkVariant()
    )
}

fun ButtonVariant.toSdkVariant(): ButtonParams.Variant {
    return when (this) {
        ButtonVariant.DARK -> ButtonParams.Variant.DARK
        else -> ButtonParams.Variant.LIGHT
    }
}

fun PromoBannerParamsData.toSdkPromoBannerParams(): PromoBannerParams {
    return PromoBannerParams.Builder()
        .build()
}

/**
 * Utility functions for data validation
 */
object RevolutPayDataValidator {
    
    fun validateCustomerData(customer: CustomerData?): Boolean {
        if (customer == null) return true
        
        // Validate email format if provided
        customer.email?.let { email ->
            if (!android.util.Patterns.EMAIL_ADDRESS.matcher(email).matches()) {
                return false
            }
        }
        
        // Validate phone format if provided
        customer.phone?.let { phone ->
            if (!phone.startsWith("+") || phone.length < 10) {
                return false
            }
        }
        
        // Validate date of birth if provided
        customer.dateOfBirth?.let { dob ->
            if (dob.day < 1 || dob.day > 31 || dob.month < 1 || dob.month > 12 || dob.year < 1900 || dob.year > 2100) {
                return false
            }
        }
        
        // Validate country if provided
        customer.country?.let { country ->
            if (country.value.length != 2 || !country.value.all { it.isUpperCase() }) {
                return false
            }
        }
        
        return true
    }
    
    fun validateButtonParams(params: ButtonParamsData): Boolean {
        // Validate currency format if provided
        params.boxTextCurrency?.let { currency ->
            if (currency.length != 3 || !currency.all { it.isUpperCase() }) {
                return false
            }
        }
        
        return true
    }
    
    fun validateSdkInitData(data: SdkInitData): Boolean {
        // Validate return URI format
        if (!data.returnUri.startsWith("http://") && !data.returnUri.startsWith("https://")) {
            return false
        }
        
        // Validate merchant public key (should be non-empty)
        if (data.merchantPublicKey.isBlank()) {
            return false
        }
        
        return true
    }
}

File: lib/android/models/revolut_pay_models.dart
import '../enums/revolut_enums.dart';

/// Customer data structure
class CustomerData {
  final String? name;
  final String? email;
  final String? phone;
  final DateOfBirthData? dateOfBirth;
  final String? country; // Now using string for country code

  const CustomerData({
    this.name,
    this.email,
    this.phone,
    this.dateOfBirth,
    this.country,
  });

  Map<String, dynamic> toMap() {
    return {
      'name': name,
      'email': email,
      'phone': phone,
      'dateOfBirth': dateOfBirth?.toMap(),
      'country': country,
    };
  }

  factory CustomerData.fromMap(Map<String, dynamic> map) {
    return CustomerData(
      name: map['name'] as String?,
      email: map['email'] as String?,
      phone: map['phone'] as String?,
      dateOfBirth: map['dateOfBirth'] != null
          ? DateOfBirthData.fromMap(map['dateOfBirth'] as Map<String, dynamic>)
          : null,
      country: map['country'] as String?,
    );
  }
}

/// Date of birth data structure
class DateOfBirthData {
  final int day;
  final int month;
  final int year;

  const DateOfBirthData({
    required this.day,
    required this.month,
    required this.year,
  });

  Map<String, dynamic> toMap() {
    return {'day': day, 'month': month, 'year': year};
  }

  factory DateOfBirthData.fromMap(Map<String, dynamic> map) {
    return DateOfBirthData(
      day: map['day'] as int,
      month: map['month'] as int,
      year: map['year'] as int,
    );
  }
}

/// Button parameters data structure
class ButtonParamsData {
  final ButtonRadius radius;
  final ButtonSize size;
  final BoxText boxText;
  final String? boxTextCurrency;
  final VariantModesData? variantModes;

  const ButtonParamsData({
    this.radius = ButtonRadius.medium,
    this.size = ButtonSize.large,
    this.boxText = BoxText.none,
    this.boxTextCurrency,
    this.variantModes,
  });

  Map<String, dynamic> toMap() {
    return {
      'radius': radius.value,
      'size': size.value,
      'boxText': boxText.value,
      'boxTextCurrency': boxTextCurrency,
      'variantModes': variantModes?.toMap(),
    };
  }

  factory ButtonParamsData.fromMap(Map<String, dynamic> map) {
    return ButtonParamsData(
      radius: ButtonRadius.values.firstWhere(
        (e) => e.value == map['radius'],
        orElse: () => ButtonRadius.medium,
      ),
      size: ButtonSize.values.firstWhere(
        (e) => e.value == map['size'],
        orElse: () => ButtonSize.large,
      ),
      boxText: BoxText.values.firstWhere(
        (e) => e.value == map['boxText'],
        orElse: () => BoxText.none,
      ),
      boxTextCurrency: map['boxTextCurrency'] as String?,
      variantModes: map['variantModes'] != null
          ? VariantModesData.fromMap(
              map['variantModes'] as Map<String, dynamic>,
            )
          : null,
    );
  }
}

/// Variant modes data structure
class VariantModesData {
  final ButtonVariant darkTheme;
  final ButtonVariant lightTheme;

  const VariantModesData({
    this.darkTheme = ButtonVariant.dark,
    this.lightTheme = ButtonVariant.light,
  });

  Map<String, dynamic> toMap() {
    return {'darkTheme': darkTheme.value, 'lightTheme': lightTheme.value};
  }

  factory VariantModesData.fromMap(Map<String, dynamic> map) {
    return VariantModesData(
      darkTheme: ButtonVariant.values.firstWhere(
        (e) => e.value == map['darkTheme'],
        orElse: () => ButtonVariant.dark,
      ),
      lightTheme: ButtonVariant.values.firstWhere(
        (e) => e.value == map['lightTheme'],
        orElse: () => ButtonVariant.light,
      ),
    );
  }
}

/// Promotional banner parameters data structure
class PromoBannerParamsData {
  final String transactionId;
  final int paymentAmount;
  final RevolutCurrency currency;
  final CustomerData? customer;

  const PromoBannerParamsData({
    required this.transactionId,
    required this.paymentAmount,
    required this.currency,
    this.customer,
  });

  Map<String, dynamic> toMap() {
    return {
      'transactionId': transactionId,
      'paymentAmount': paymentAmount,
      'currency': currency.value,
      'customer': customer?.toMap(),
    };
  }

  factory PromoBannerParamsData.fromMap(Map<String, dynamic> map) {
    return PromoBannerParamsData(
      transactionId: map['transactionId'] as String,
      paymentAmount: map['paymentAmount'] as int,
      currency: RevolutCurrency.values.firstWhere(
        (e) => e.value == map['currency'],
        orElse: () => RevolutCurrency.gbp,
      ),
      customer: map['customer'] != null
          ? CustomerData.fromMap(map['customer'] as Map<String, dynamic>)
          : null,
    );
  }
}

/// Order result callback data
class OrderResultData {
  final bool success;
  final String? orderId;
  final String? error;
  final String? cause;

  const OrderResultData({
    required this.success,
    this.orderId,
    this.error,
    this.cause,
  });

  Map<String, dynamic> toMap() {
    return {
      'success': success,
      'orderId': orderId,
      'error': error,
      'cause': cause,
    };
  }

  factory OrderResultData.fromMap(Map<String, dynamic> map) {
    return OrderResultData(
      success: map['success'] as bool,
      orderId: map['orderId'] as String?,
      error: map['error'] as String?,
      cause: map['cause'] as String?,
    );
  }
}

/// Controller creation result data
class ControllerResultData {
  final String controllerId;
  final bool success;

  const ControllerResultData({
    required this.controllerId,
    required this.success,
  });

  Map<String, dynamic> toMap() {
    return {'controllerId': controllerId, 'success': success};
  }

  factory ControllerResultData.fromMap(Map<String, dynamic> map) {
    return ControllerResultData(
      controllerId: map['controllerId'] as String,
      success: map['success'] as bool,
    );
  }
}

/// Button creation result data
class ButtonResultData {
  final String buttonId;
  final bool success;

  const ButtonResultData({required this.buttonId, required this.success});

  Map<String, dynamic> toMap() {
    return {'buttonId': buttonId, 'success': success};
  }

  factory ButtonResultData.fromMap(Map<String, dynamic> map) {
    return ButtonResultData(
      buttonId: map['buttonId'] as String,
      success: map['success'] as bool,
    );
  }
}

/// Banner creation result data
class BannerResultData {
  final String bannerId;
  final bool success;

  const BannerResultData({required this.bannerId, required this.success});

  Map<String, dynamic> toMap() {
    return {'bannerId': bannerId, 'success': success};
  }

  factory BannerResultData.fromMap(Map<String, dynamic> map) {
    return BannerResultData(
      bannerId: map['bannerId'] as String,
      success: map['success'] as bool,
    );
  }
}

/// Payment flow data
class PaymentFlowData {
  final String orderToken;
  final bool savePaymentMethodForMerchant;

  const PaymentFlowData({
    required this.orderToken,
    this.savePaymentMethodForMerchant = false,
  });

  Map<String, dynamic> toMap() {
    return {
      'orderToken': orderToken,
      'savePaymentMethodForMerchant': savePaymentMethodForMerchant,
    };
  }

  factory PaymentFlowData.fromMap(Map<String, dynamic> map) {
    return PaymentFlowData(
      orderToken: map['orderToken'] as String,
      savePaymentMethodForMerchant:
          map['savePaymentMethodForMerchant'] as bool? ?? false,
    );
  }
}

/// SDK initialization data
class SdkInitData {
  final RevolutEnvironment environment;
  final String returnUri;
  final String merchantPublicKey;
  final bool requestShipping;
  final CustomerData? customer;

  const SdkInitData({
    required this.environment,
    required this.returnUri,
    required this.merchantPublicKey,
    this.requestShipping = false,
    this.customer,
  });

  Map<String, dynamic> toMap() {
    return {
      'environment': environment.value,
      'returnUri': returnUri,
      'merchantPublicKey': merchantPublicKey,
      'requestShipping': requestShipping,
      'customer': customer?.toMap(),
    };
  }

  factory SdkInitData.fromMap(Map<String, dynamic> map) {
    return SdkInitData(
      environment: RevolutEnvironment.values.firstWhere(
        (e) => e.value == map['environment'],
        orElse: () => RevolutEnvironment.sandbox,
      ),
      returnUri: map['returnUri'] as String,
      merchantPublicKey: map['merchantPublicKey'] as String,
      requestShipping: map['requestShipping'] as bool? ?? false,
      customer: map['customer'] != null
          ? CustomerData.fromMap(map['customer'] as Map<String, dynamic>)
          : null,
    );
  }
}
